### PR TITLE
Add slurm multi-process auto bootstrapping interface

### PR DESCRIPTION
**Pull request by nvcastet**
* Generic bootstrapping interface to plug in new environments
* Slurm bootstrapping

### CODE REVIEW

**Review by skye — CHANGES_REQUESTED** [review]
This is neat! Overall I like the idea, but I have some thoughts on the overall structure and some details of the Cluster API. Lemme know if you wanna discuss any of this over video or chat, feel free to email or Google chat me.

File: jax/_src/clusters/cluster_init.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+      return coordinator_address, num_processes, process_id
+
+    @classmethod
+    @abstractmethod
+    def is_env_present(cls):
+      ...
+
+    @classmethod
+    @abstractmethod
+    def get_coordinator_address(cls):
```
[the remark this answers is not in the document]
Comment (nvcastet) [reply jax/_src/clusters/cluster_init.py:35]: Yes good point, I can add some doc here.

File: jax/_src/clusters/__init__.py [hunk 2]
```diff
@@ -0,0 +1,3 @@
+from .cluster_init import Cluster
+from .cloud_tpu_init import TpuCluster
+from .slurm_init import SlurmCluster
```
[the remark this answers is not in the document]
Comment (nvcastet) [reply jax/_src/clusters/__init__.py:3]: Correct, the order will dictate a priority if several environments are available. I will add a comment.

File: jax/_src/clusters/slurm_init.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+_LOCAL_PROCESS_ID = 'SLURM_LOCALID'
+
+class SlurmCluster(Cluster):
+  @classmethod
+  def is_env_present(cls):
+    return _JOBID_PARAM in os.environ
+
+  @classmethod
+  def get_coordinator_address(cls):
+    # Pick port in ephemeral range [(65535 - 2^3 + 1), 65535]
```
[the remark this answers is not in the document]
Comment (nvcastet) [reply jax/_src/clusters/slurm_init.py:17]: Good catch, I meant to have a bigger range like the last `2^12` ports of the ephemeral range but forgot to update it. Yes you cannot use `/proc/sys/net/ipv4/ip_local_port_range` since the logic to get the port needs to lead to the exact same port on all the processes to be able to connect to the open port of the 1st process. Using `/proc/sys/net/ipv4/ip_local_port_range` may give you different answer on different nodes.

File: jax/_src/clusters/cluster_init.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+      ...
+
+    @classmethod
+    @abstractmethod
+    def default_to_single_process_per_device(cls):
+      ...
+
+    @staticmethod
+    def is_single_process_per_device(env):
+      user_var = os.environ.get('JAX_SINGLE_PROCESS_PER_DEVICE', None)
```
[the remark this answers is not in the document]
Comment (nvcastet) [reply jax/_src/clusters/cluster_init.py:60]: @skye Correct, we already know the number of total processes of the job and also the number of processes per node.

The only case which is an issue is when the cluster tells you there are 1 process and 1 node. How do we know if the user wants to run on a single process with one device, or with all the devices, or in-between. We could make it flexible as you mentioned and have the user set `JAX_VISIBLE_DEVICES`. But what would be a good default? For GPUs, we would like to default to 1 but for other cases to all.
We don't need to worry about `CUDA_VISIBLE_DEVICES`, this is a CUDA driver setting, so the CUDA API that XLA is using will honor `CUDA_VISIBLE_DEVICES`.

**Comment by nvcastet** [comment]
@skye I rebased the code and added a commit to address your comments. Let me know what you think.

File: jax/_src/clusters/slurm_cluster.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    return _JOBID_PARAM in os.environ
+
+  @classmethod
+  def get_coordinator_address(cls):
+    # Pick port in ephemeral range [(65535 - 2^12 + 1), 65535]
+    port = int(os.environ[_JOBID_PARAM]) % 2**12 + (65535 - 2**12 + 1)
+
+    # Parse the first hostname of the job
+    # node_list potential formats are 'node1', 'node1,host2',
+    # 'node[001-0015],host2', and 'node[001,007-015],host2'.
```
Comment (skye) [inline jax/_src/clusters/slurm_cluster.py:23]: I'm confused what `node` and `host` mean, can you also comment which part is the hostname we're looking for?

File: jax/_src/clusters/cluster.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+  """
+
+  _cluster_types: List['ClusterEnv'] = []
+
+  def __init_subclass__(cls, **kwargs):
+    super().__init_subclass__(**kwargs)
+    cls._cluster_types.append(cls)
+
+  @classmethod
+  def auto_detect_unset_distributed_params(cls, coordinator_address,
```
Comment (skye) [inline jax/_src/clusters/cluster.py:22]: Can you put types on the params and outputs for all of these interface methods?

File: jax/_src/clusters/cluster.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+  @classmethod
+  def get_process_count(cls):
+    raise NotImplementedError
+
+  @classmethod
+  def get_process_id(cls):
+    raise NotImplementedError
+
+  @classmethod
+  def get_local_process_id(cls):
```
Comment (skye) [inline jax/_src/clusters/cluster.py:77]: This can apparently return None, so I think it needs a docstring explaining what that means. And should this return None by default instead of raising NotImplemented?

File: jax/_src/clusters/cluster.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    """
+    raise NotImplementedError
+
+  @classmethod
+  def get_process_count(cls):
+    raise NotImplementedError
+
+  @classmethod
+  def get_process_id(cls):
+    raise NotImplementedError
```
Comment (skye) [inline jax/_src/clusters/cluster.py:74]: I think `get_coordinator_address`, `get_process_count`, and `get_process_id` should be abstract methods (https://docs.python.org/3/library/abc.html#abc.abstractmethod)

**Review by skye — CHANGES_REQUESTED** [review]
It's looking good! Most of the comments are just to make the code clearer for future readers + users

File: jax/_src/clusters/cluster.py (hunk 7 above)
Comment (nvcastet) [reply jax/_src/clusters/cluster.py:77]: It cannot be None. To make sure it is dealt with by the subclass, it would let the exception. I can add a docstring on what it means.

File: jax/_src/clusters/cluster.py (hunk 8 above)
Comment (nvcastet) [reply jax/_src/clusters/cluster.py:74]: You may have seem that I had them as abstract methods in the first version of the PR. By trying it and reading the doc carefully, abstract method decorator in python will only help if you are instantiating subclass objects. In our case we are never creating objects, we just iterate over the subclass types.

File: docs/multi_process.md [hunk 9]
```diff
@@ -64,7 +64,11 @@ The API {func}`jax.distributed.initialize` takes several arguments, namely:
     cluster will connect.
   * `num_processes`: the number of processes in the cluster
   * `process_id`: the ID number of this process, in the range `[0 ..
-  num_processes)`.
+    num_processes)`.
+  * `local_device_ids`: Restricts the visible devices of the current process to
+    ``local_device_ids``. Defaults to all local devices being visible to the
+    process except when processes are launched via Slurm on GPUs.
+    In that case, it will default to a single device per process.
```
Comment (skye) [inline docs/multi_process.md:71]: Let's actually remove the whole "Defaults to ... single deivce per process." section, since you explain that immediately below. (Sorry if I'm contradicting myself, I just looked at the rendered docs and thought it was a little awkward to read it twice in short succession.)

**Review by skye — CHANGES_REQUESTED** [review]
Looks great! One more small round of changes and then let's get this in.

File: jax/_src/distributed.py [hunk 10]
```diff
@@ -33,24 +33,18 @@ class State:
   def initialize(self,
                  coordinator_address: Optional[str] = None,
                  num_processes: Optional[int] = None,
-                 process_id: Optional[int] = None):
+                 process_id: Optional[int] = None,
+                 local_device_ids: Union[int, Sequence[int], None] = None):
```
[the remark this answers is not in the document]
Comment (nvcastet) [reply jax/_src/distributed.py:37]: oh, I did not know that was possible. Thanks for sharing.

File: jax/_src/clusters/cluster.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+    cls._cluster_types.append(cls)
+
+  @classmethod
+  def auto_detect_unset_distributed_params(cls,
+                                           coordinator_address: Optional[str],
+                                           num_processes: Optional[int],
+                                           process_id: Optional[int],
+                                           local_device_ids: Optional[Union[int, Sequence[int]]]
+                                          ) -> Tuple[Optional[str], Optional[int], Optional[int],
+                                                     Optional[Union[int, Sequence[int]]]]:
```
Comment (skye) [inline jax/_src/clusters/cluster.py:29]:
```suggestion
                                           local_device_ids: Optional[Sequence[int]]
                                          ) -> Tuple[str, int, int, 
                                                     Optional[Sequence[int]]]:
```

Sorry I missed this earlier. These should definitely be set (except local device ids), and we normalize the local devices to a list in distributed.init right?

Comment (skye) [reply jax/_src/clusters/cluster.py:29]: Oh nevermind, they're None if we can't find an environment. I think the local_device_ids should always be a list and never an int though, for both the parameter and returned value.

Comment (nvcastet) [reply jax/_src/clusters/cluster.py:29]: correct, yes it should be none if we don't have an environment. Let me fix the `local_device_ids` type.

**Review by skye — APPROVED** [review]

**Comment by skye** [comment]
Oh, you need to add the copyright boilerplate to the top of all new files. Copy/paste from any existing file and change the copyright year to 2022.

File: jax/_src/clusters/cluster.py [hunk 12]
```diff
@@ -0,0 +1,108 @@
+# Copyright (c) 2022, NVIDIA CORPORATION. All rights reserved.
```
Comment (yashk2810) [inline jax/_src/clusters/cluster.py:1]: @skye should this be NVIDIA corporation or Google LLC?

Comment (yashk2810) [reply jax/_src/clusters/cluster.py:1]: @hawkinsp

Comment (hawkinsp) [reply jax/_src/clusters/cluster.py:1]: The rules we follow here are:
https://opensource.google/documentation/reference/releasing/contributions#copyright

(I think we'd be very open to changing the copyright attribution to "The JAX authors" but we'd have to do that for the entire project.)

Comment (nvcastet) [reply jax/_src/clusters/cluster.py]: Sounds good, I just pushed a commit to follow your guideline. But yes if we don't want individual company copyright, it would make sense to switch to "The JAX authors" to match the legal reality of the files copyright.

**Review by skye — APPROVED** [review]

**Review by skye — APPROVED** [review]

**Review by skye — APPROVED** [review]
